Master port #50773: zypper's locks to holds

oeuftete opened the pull request:
Port of #50773.

----

### What does this PR do?

Deprecates `add_lock` and `remove_lock` in favour of `hold` and `unhold` functions.

### What issues does this PR fix or reference?

Fixes: #56922 (as part of making the original PR work)

### Merge requirements satisfied?

- [x] Docs
- [x] Changelog
- [x] Tests written/updated*: There are no new tests, but moving the functionality here to `hold`/`unhold` activated a handful of package integration tests (which flushed out some bugs)

### Commits signed with GPG?

No

Commit:
Add hold function

Commit:
Add unhold function

Commit:
Add warnings

Commit:
Fix variously broken hold/lock zypper functions

oeuftete commented:
I wanted to get this master port up as part of addressing #56922 and customer support issues related to the `hold`/`unhold` functionality, which is already part of the SuSE-provided Salt packages.  I can not currently take responsibility for writing the required test cases.  @saltstack/team-suse, maybe?

brejoc commented:
I'm missing a bit of context here. Seems like #50773 was originally merged but then reverted again!? Was it incomplete?

brejoc commented:
Ah, sorry! I missed that that was `develop` for #50773. Okay, got it now.

ScriptAutomate commented:
The functions being deprecated here should include something similar to the following at the top of their respective function docstrings:

```rst
.. deprecated:: 3002,3001.1
    This function is deprecated. Please use ``unhold()`` / ``hold()`` instead.
```

We may need to update docs to suggest `.. deprecation::` Sphinx / rst directives here, too:
- https://docs.saltstack.com/en/latest/topics/development/deprecations.html
  - Source: https://github.com/saltstack/salt/blob/master/doc/topics/development/deprecations.rst

Commit:
Run pre-commit against changed files

Commit:
Fix result statuses for zypper un/hold

Commit:
Clean up some split strings in yumpkg

Commit:
Add versionadded for zypperpkg un/hold

Commit:
Allow both name and pkgs for zypper un/hold

This is expected by the pkg.installed state, and is consistent with
other package states.

Commit:
Add Suse support to installed-hold integration test

Commit:
Add pkg state doc for zypper hold support

Commit:
Run pre-commit against changed files

ScriptAutomate commented:
@oeuftete Can you resolve the conflicts in `salt/modules/zypperpkg.py`? I'd like to get this in front of some reviewers to see how close this is to getting merged.

twangboy commented on salt/modules/win_pkg.py:
```diff
@@ -2310,15 +2298,7 @@ def _get_name_map(saltenv="base"):
-    for k in name_map:
-        u_name_map[k] = name_map[k]
-    return u_name_map
+    return get_repo_data(saltenv).get("name_map", {})
```
We still need to create a reverse map here. Should be:
```
    u_name_map = {}
    name_map = get_repo_data(saltenv).get("name_map", {})
    for k in name_map:
        u_name_map[k] = name_map[k]
    return u_name_map
```

oeuftete replied:
Good catch.  Even from my commit comment, I'm not 100% sure why I had to mess with this file four months ago (and then messed it up), but I believe it was to satisfy pre-commit requirements at the time. This file has no business being changed in this PR if it doesn't need to, so I'm going try to just rip the commit changing this out.

Interesting that no existing tests caught my mistake, this might be something to address.

oeuftete replied:
If it's meant to return the reverse map, I believe it was broken before...

```diff
-    u_name_map = {}
-    name_map = get_repo_data(saltenv).get("name_map", {})
-
-    if not six.PY2:
-        return name_map
-
-    for k in name_map:
-        u_name_map[k] = name_map[k]
-    return u_name_map
+    return get_repo_data(saltenv).get("name_map", {})
```

The reverse map would never be returned under py3.  So that's why this was taken out when I de-sixed it.

oeuftete replied:
@twangboy It looks like the original change is no longer required (the Windows tests pass), so I don't need to touch this module in this PR.  If the lack of the reverse map is a bug, that should be addressed in a separate PR.

oeuftete reviewed (commented):
https://jenkins.saltproject.io/job/pr-opensuse15-py3-pytest/job/PR-56935/4/testReport/tests.integration.states.test_pkg/PkgTest/test_pkg_017_installed_held_equals_false/

`tests.integration.states.test_pkg.PkgTest.test_pkg_017_installed_held_equals_false` is now failing, which is related to the changes here.

```
E           AssertionError: 'held' not found in 'The following packages were installed/updated: lynx'
```

Commit:
Update test from #58883 to work with new work

oeuftete commented:
Ready for review again.  Thanks to @garethgreenaway for helping sort out the latest test failure.

OrangeDog commented:
Why were these renamed in the first place? The other two lock functions (`clean_locks` and `list_locks`) were not renamed.

Zypper has no concept of "hold". It calls them package locks.
